kani-driver/src/main.rs

```diff
@@ -61,30 +59,20 @@ fn cargokani_main(input_args: Vec<OsString>) -> Result<()> {
 
     let metadata = ctx.collect_kani_metadata(&outputs.metadata)?;
     let harnesses = ctx.determine_targets(&metadata)?;
-    let sorted_harnesses = metadata::sort_harnesses_by_loc(&harnesses);
     let report_base = ctx.args.target_dir.clone().unwrap_or(PathBuf::from("target"));
 
-    let mut failed_harnesses: Vec<&HarnessMetadata> = Vec::new();
-
-    for harness in &sorted_harnesses {
-        let harness_filename = harness.pretty_name.replace("::", "-");
-        let report_dir = report_base.join(format!("report-{}", harness_filename));
-        let specialized_obj = outputs.outdir.join(format!("cbmc-for-{}.out", harness_filename));
-        ctx.run_goto_instrument(
-            &linked_obj,
-            &specialized_obj,
-            &outputs.symtabs,
-            &harness.mangled_name,
-        )?;
+    let runner = harness_runner::HarnessRunnerSession {
+        sess: &ctx,
+        linked_obj: &linked_obj,
+        report_base: &report_base,
+        symtabs: &outputs.symtabs,
+        retain_specialized_harnesses: true,
+    };
 
-        let result = ctx.check_harness(&specialized_obj, &report_dir, harness)?;
-        if result == VerificationStatus::Failure {
-            failed_harnesses.push(harness);
-        }
-    }
+    let results = runner.check_all_harnesses(&harnesses)?;
 
-    ctx.inform_if_no_failed(&failed_harnesses);
-    ctx.print_final_summary(&sorted_harnesses, &failed_harnesses)
+    ctx.inform_if_no_failed(&results);
```

Comment:
Still not clear to me why this is part of `KaniSession` and not `HarnessRunner`.

Reply:
Actually, is there a reason to call these two functions here instead of putting them inside `check_all_harnesses`?

Reply:
`print_final_results` consumes the kani session (if there were failures it needs to drop it, and then exit the process with appropriate error code.)

So that's why that needs to be a separate function called by main. `assess` for instance, will do other things after a `HarnessRunner` completes.

As for `check_harness`, it does not use any of the data in `HarnessRunner`, and that's why it's not a function under that structure. Every argument it gets, for instance, is specialized per-harness, not something that's run-wide. (`report_dir` for instance is not `report_base`).